typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+
+Semantically, upcalls have two arguments: the userdata pointer (which was passed
+to Subscribe), and an `ArbitraryData(4)` of data from the syscall driver. This
+allows syscall drivers to pass up to three values to each invoked upcall. Upcall
+functions must have an ABI compatible with the following signature (specified in
+both Rust and C):
+
+```C
+struct UpcallArbitraryData {
+    Register registers[4];
+}
+
+typedef void(void*, UpcallArbitraryData) UpcallFn;
+```
+
+```Rust
+#[repr(C)]
+struct UpcallArbitraryData {
+    registers: [Register; 4],
+}
+
+type UpcallFn = unsafe extern "C" fn(*mut (), UpcallArbitraryData);
+```
+
+### Yield
+
+Arguments:
+
+| Yield type | a1  | a2            | a3               |
+| ---------- | --- | ------------- | ---------------- |
+| no-wait    | `0` | Unused        | Unused           |
+| wait       | `1` | Unused        | Unused           |
+| wait-for   | `2` | Driver number | Subscribe number |
+
+If there is no pending upcall, yield-no-wait will return `()`.
+
+If there is an upcall to invoke, yield-wait and yield-no-wait return `(upcall fn
+pointer, userdata: *mut (), Register, Register, Register, Register)`. The
+userspace Yield function should then pack the register values into an
+`UpcallArbitraryData` and invoke the function pointer with the userdata pointer
+and `UpcallArbitraryData` as parameters.
+
+If there is an upcall, yield-wait-for returns `(Register, Register, Register,
+Register)`. The four `Register` values are the upcall's `ArbitraryData(4)`
+parameters.
+
+### Subscribe
+
+Arguments:
+
+| Register | Argument         | Type               |
+| -------- | ---------------- | ------------------ |
+| a1       | Driver number    | `u32`              |
+| a2       | Subscribe number | `u32`              |
+| a3       | Upcall pointer   | Upcall fn pointer  |
+| a4       | Application data | `*mut ()`/`void *` |
+
+Return variants:
+
+| Outcome | Type                                                 |
+| ------- | ---------------------------------------------------- |
+| Failure | (`ErrorCode`, upcall fn pointer, `*mut ()`/`void *`) |
+| Success | (upcall fn pointer, `*mut ()`/`void *`)              |
+
+### Command
+
+Each specific Command instance (combination of Driver and Command number) must
+specify its argument type, failure type, and success type. If a Command instance
+is invoked with the wrong argument type, the Command call will failure with
+error code `INVALID`.
+
+Arguments:
+
+| Register | Argument                    | Type            |
+| -------- | --------------------------- | --------------- |
+| a1       | Driver number               | `u32`           |
+| a2       | Command number              | `u32`           |
+| a3-...   | Instance-specific arguments | `ArbitraryData` |
+
+The return variants are specific to the Command instance, but MUST fit into an
+`ArbitraryData(15)`. This guarantees that Command's return value will clobber no
+more than 15 registers.
```

Comment:
But isn't `ArbitraryData(15)` bigger than the number of registers ARM has available for returns?

Reply:
Also, where did the 15 registers requirement come from?

Reply:
> But isn't `ArbitraryData(15)` bigger than the number of registers ARM has available for returns?

My intent was that syscall drivers that use more than 11 registers would be unavailable on ARM, but I never stated that directly.

> Also, where did the 15 registers requirement come from?

The reasoning was buried deeply in the Properties section, but it belongs here, so I replaced this whole paragraph with the following:

> The return variants are specific to the Command instance. On ARM, they MUST fit into an `ArbitraryData(11)`. On RISC-V, they MUST fit into an `ArbitraryData(15)`.
>
> Design reasoning: ARM has instructions to push and pop multiple registers, so clobbering all 11 registers is relatively inexpensive. On RISC-V, this clobbers only caller-saved registers, so the compiler should have the context it needs to only save registers that are currently in use.

and removed the now-redundant explanation from the Properties section.